### PR TITLE
You can't use beds when there are hostile mobs nearby

### LINKED ISSUE #705: You can use beds when monsters are nearby
You can sleep in a bed although there are monsters nearby. 
Normally, when a monster is in range of 8 horizontally and 5 vertically you can't use a bed and you get the message "You may not rest now, there are monsters nearby."

### PR DESCRIPTION
**Pull request by Bond-009**
Fixes #705

### CODE REVIEW

**Comment by Seadragon91** [comment]
I think the detection for the height needs to be adjusted. In cBoundingBox-API I read that the height starts from minY to maxY + height. A monster placed nearby a bed with a height diff of 3 blocks deep is not detected.

**Comment by Seadragon91** [comment]
Currently it's minY to maxY + 5, should be minY - 5 to maxY + 5 or?

**Comment by Bond-009** [comment]
So the code in cBoundingBox needs to be changed?

**Comment by spekdrum** [comment]
I introduced an overloaded constructor for cBoundingBox in PR #3798 that allows to do that:
```
cBoundingBox::cBoundingBox(const Vector3d & a_Pos, double a_Radius, double a_Height, double a_VerticalOffset) :
 	m_Min(a_Pos.x - a_Radius, a_Pos.y + a_VerticalOffset, a_Pos.z - a_Radius),
+	m_Max(a_Pos.x + a_Radius, a_Pos.y + a_VerticalOffset + a_Height, a_Pos.z + a_Radius)
 {
 }
```

Maybe you can copy it, but it's not revised yet.

**Comment by Seadragon91** [comment]
@Bond-009 cBoundingBox is correct and the API also. You could change it to this:
`a_Player->GetPosition()`
to
`a_Player->GetPosition() - Vector3i(0, 5, 0)`
This will subtract 5 blocks from the start pos and then it's okay.

**Comment by Bond-009** [comment]
Ok, I wasn't sure if that was intended :)

File: src/Blocks/BlockBed.cpp [hunk 1]
```diff
@@ -88,14 +107,23 @@ bool cBlockBedHandler::OnUse(cChunkInterface & a_ChunkInterface, cWorldInterface
...
+		if ((Meta & 0x4) == 0x4)
+		{
+			a_Player->SendMessageFailure("This bed is occupied");
+		}
+		else
 		{
-			NIBBLETYPE Meta = a_ChunkInterface.GetBlockMeta(a_BlockX, a_BlockY, a_BlockZ);
-			if ((Meta & 0x4) == 0x4)
+			cFindMobs FindMobs;
+			if (!a_Player->GetWorld()->ForEachEntityInBox(cBoundingBox(a_Player->GetPosition() - Vector3i(0, 5, 0), 8, 10), FindMobs))
```
Comment (Seadragon91) [inline src/Blocks/BlockBed.cpp]: Why changed `8, 5` to `8, 10`? Now it its startY - 5 to startY + 10

Comment (Bond-009) [inline src/Blocks/BlockBed.cpp]: Yes, that should be a 5 blocks above and under the player, no?

Comment (Seadragon91) [inline src/Blocks/BlockBed.cpp]: Ah ops, sorry :) Really confused me hehe

**Comment by Seadragon91** [comment]
For closing a issue via a [keyword](https://help.github.com/articles/closing-issues-via-commit-messages/#keywords-for-closing-issues) it has to be followed by the number.
Change it to `Fixes #705`

**Comment by Bond-009** [comment]
Cool, I didn't know that :+1:

**Pull request merged**